Adds a utility Covariance class for NDData

kbwestfall opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Addresses #16648 .

This is a basic port of a nearly identical class from the MaNGA data-analysis pipeline.  This is a basic utility class that allows for the construction and storage of covariance matrices.  The code uses `scipy.sparse` to only store the upper triangle of non-zero values.  When written to a fits file, only these values are stored to minimize file size.

I'm issuing this draft PR so others can see the code, read some of the documentation, and provide comments.  At this point, I haven't fully integrated the class into the NDData module.  I've isolated the additions to the new files.  This is my first astropy PR, so please let me know if I've missed any important dev/PR protocols.  In particular, please advise on how I should enable scipy to be an optional dependency.

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

pllim commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1034 @@
+
+    Parameters
+    ----------
+    inp : `scipy.sparse.csr_matrix`, optional
```
`scipy` is still an "optional" dependency for `astropy`. Does this mean anything touching the stuff from this module requires `scipy` to run?

pllim replied:
Please put your thoughts here:

* #12448

dhomeier replied:
See my review comments; I think if it is generally only available with scipy, that's acceptable.
Whether it would be worth offering a stripped-down version with dense numpy arrays depends on the extra work involved with that, and if you'd still consider such an option usable given the performance impact.
There are certainly plenty of use cases with array lengths of only ~100 or less where it probably would not hurt, as in the original `specutils` example where this came up. But then `specutils` already depends on scipy, so would never need to use that alternative version.

pllim commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1034 @@
+        quiet=False,
+    ):
+        r"""
+        Read covariance data from a fits file.
```
FITS is acronym, so needs to be capitalize. This comment applies to all similar occurences.

```suggestion
        Read covariance data from a FITS file.
```

kbwestfall replied:
Fixed!

pllim commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1034 @@
+        )
+
+    @classmethod
+    def from_fits(
```
This method relies too much on special FITS header keywords. If these keywords only apply to, say, one very specific telescope or instrument, it should not be in core `astropy` library.

kbwestfall replied:
Are you specifically referring to the `INDX*` and `RHO*` keywords?  Yes, these are specific to this class, but they are not specific to any instrument or telescope.  They are consistent with general nomenclature used for covariance matrices.  Although, I should probably generalize the `I` and `J` to something like `0` and `1`...

pllim replied:
So, these are formalized in some WCS papers or something? (I never read them LoL)

kbwestfall replied:
No, there's no formal WCS for these (that I know of).

pllim commented on docs/nddata/covariance.rst:
Does this new page needs to be in a TOC somewhere?

pllim replied:
As for the heading level, please refer to other similar pages in `docs/nddata` to be consistent.

kbwestfall replied:
Yes, it should be, but I haven't fully integrated the code and docs yet.  I wanted to wait until I got some confirmation that adding this `Covariance` class was going to go through before doing that.

For the heading levels, I tried to follow `docs/nddata/nddata.rst`.

pllim commented on docs/nddata/covariance.rst:
```diff
@@ -0,0 +1,187 @@
+    distribution::
+
+    >>> import numpy as np
+    >>> from mangadap.util.covariance import Covariance
```
Is `mangadap` a new package? I don't think I want that as doc build dependency in astropy. Are you able to build an example without this specialized package?

kbwestfall replied:
Sorry!  I missed editing this to import from astropy instead.  Fixed.

pllim commented:
Thank you. I left some preliminary comments. As mentioned in #16648 , we need to identify what truly belongs in core `astropy` library, and what is more appropriate downstream in the ecosystem.

Given how heavily this implementation relies on `scipy.sparse`, do you think some of the math can be upstreamed to `scipy`? Surely covariance does not exist only in astronomy?

kbwestfall commented on docs/nddata/covariance.rst:
```diff
@@ -0,0 +1,187 @@
+
+>>> import numpy as np
+>>> from mangadap.util.covariance import Covariance
+>>> ofile = 'test_covar_io.fits'
```
Yeah, I was worried about this.  Note this is identical to one of the unit tests.  Can you advise on how to adjust the I/O tests?

pllim replied:
We can figure this out later. Might be able to reuse a test file you will add in the future or something.

dhomeier replied:
The only failure remaining seems to be on the `hdu.info()` below – looks like you are assuming a different info level in the example than what is set in the doctest.

kbwestfall replied:
Doctests should pass now.

kbwestfall reviewed (commented):
Thanks for the quick comments, @pllim !

kbwestfall commented:
> Given how heavily this implementation relies on `scipy.sparse`, do you think some of the math can be upstreamed to `scipy`? Surely covariance does not exist only in astronomy?

Embarrassingly, I didn't realize there is a [scipy.stats.Covariance](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.Covariance.html) class, but it's purpose doesn't overlap much with this class.  The scipy implementation provides a representation of the covariance matrix, whereas this class (and the request behind #16648) is largely about construction, access, and storage.

I don't have a good sense as to whether this is better upstreamed to scipy, but I imagine things like `astropy.nddata.NDUncertainty` exist in astropy for reasons similar to why it's reasonable to include this Covariance class.  Of course, I'm interested to hear opinions.

pllim commented:
> construction, access, and storage

Would it make sense to subclass `scipy.stats.Covariance`? 💭

dhomeier commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1034 @@
+        ivar_hdu = fits.ImageHDU(data=np.ma.power(var, -1.0).filled(0.0), name="IVAR")
+        return _hdr, ivar_hdu, covar_hdu
+
+    def write(self, ofile, reshape=False, hdr=None, overwrite=False):
```
This will perhaps need most discussion – I think `NDData` objects, or rather their subclasses, are generally (expected to be) using the unified IO framework, and having a class bring in its own `write` method will probably not fit in with that. More specifically, #16648 envisions `Covariance` to be associated with an `NDData` object, to be read and written together.
I _think_ this means such subclassed objects would set up their own readers and writers and register them, as it is done e.g. for `CCDData` and would be done for `Spectrum1D`, and handle themselves exactly how and where to store a `covariance` attribute, just as it is done for `uncertainty`. The `output_hdus` method defined above would be crucial for that, as it is arguably a bit more complex than just storing an uncertainty array as table row or data field.

kbwestfall replied:
I restructured this a bit.  For how I'm planning to use this in `specutils`, I just needed a to/from tables routines.  I've now implemented these, and the write/from_fits functions largely just wrap these functions.  If these FITS I/O functions are an issue, we can just remove them.

The bigger issue may be that I don't know how NDData interfaces with the unified I/O framework.  I'm now subclassing `Covariance` from `NDUncertainty`, but this may play havoc with `NDData` I/O if the `uncertainty` attribute is a `Covariance` instance.

mwcraig replied:
This has been resolved by passing the IO off to Table.

dhomeier reviewed (commented):
Thanks for your contribution; I have made some follow-up comments on how to handle the conditional scipy and matplotlib dependencies. With scipy optional that might mean the whole class would simply not be available otherwise, just like significant parts of `modeling` are.
Apart from that the next thing to do would probably be adding this as a `covariance` property to `NDData` analogously to `uncertainty`, so it would be immediately available to `NDCube` or `Spectrum1D`, and then think about how to handle it within unified IO.
This will certainly need input from maintainers or others more familiar with that framework.

kbwestfall commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1034 @@
+from pathlib import Path
+
+import numpy as np
+from matplotlib import pyplot as plt
+from scipy import sparse
+
+from astropy import log
+from astropy.io import fits
```
Can you say more about this, or point me to an example file?  I tried just adopting this edit and the code complains about `sparse` not being available.  I expect I'm missing something obvious.

dhomeier replied:
The not quite so obvious complement is to add the conditional import just around the parts of the code using `scipy.sparse`, as in
https://github.com/astropy/astropy/blob/2c17013da0e5a03f89f8bf51aa06e7d8445850d9/astropy/convolution/convolve.py#L58-L60
but since in the current implementation I guess this would be the entire `Covariance` class, need to think of a graceful way to just define that as a `NotImplemented` or similar otherwise.

kbwestfall replied:
Got it.  I've implemented something similar to what's in `astropy/cosmology/flrw/lambdacdm.py`.  Let me know what you think.

kbwestfall commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1034 @@
+
+__all__ = ["Covariance"]
+
+
```
Done.

kbwestfall commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1034 @@
+        log10 : :obj:`bool`, optional
+            Plot the base-10 log of the covariance value.
+        """
+        # Convert the covariance matrix to an array
```
I've removed this function for now.  It was mostly a convenience.

kbwestfall commented:
Hi @pllim , @dhomeier .  I'm trying to fix the doc failures, but I'm struggling with the correct syntax for references to class attributes, functions etc.  I've tried a few things, but I continue to get errors like:
`/Users/westfall/Work/packages/astropy/astropy/nddata/covariance.py:docstring of astropy.nddata.covariance.Covariance.coordinate_data:26: WARNING: py:obj reference target not found: raw_shape [ref.obj]`

I looked at a few other astropy docs and docstrings, and it's not clear to me what I'm doing wrong.  Any advice or references I should look at?  Thanks.

pllim commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1207 @@
+        -------
+        tuple
+            Two tuples providing the indices in the associate data array.  If
+            :attr:`raw_shape` is not defined, the input arrays are simply
```
Try this?
```suggestion
            ``raw_shape`` is not defined, the input arrays are simply
```

pllim commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1207 @@
+        tuple
+            A tuple of two arrays providing the indices in the covariance matrix
+            associated with the provided data array coordinates.  If
+            :attr:`raw_shape` is not defined, the input arrays are simply
```
```suggestion
            ``raw_shape`` is not defined, the input arrays are simply
```

pllim replied:
And everywhere else this is happening.

kbwestfall replied:
Thanks.  I've implemented this change here and elsewhere.

pllim commented:
Have to make sure CI does not fail when scipy is not installed.

```
astropy/nddata/tests/test_covariance.py:4: in <module>
    from scipy import sparse
E   ModuleNotFoundError: No module named 'scipy'
```

Also, have to rebase to pick up other fixes. At least RTD is green (at the time of writing this).

kbwestfall commented:
Hi @pllim 

I think I've fixed the scipy errors.  There are some lingering doctest errors on windows; see [this test](https://github.com/astropy/astropy/actions/runs/9999203872/job/27639760554?pr=16690).  Is there a way to deal with these errors that may be driven by system-dependent differences?

Also, I'm confused by the differences now shown in this PR changes, and I'm worried I somehow screwed up the rebase.  Weirdly, [this comparison](https://github.com/astropy/astropy/compare/astropy:astropy:main...kbwestfall:astropy:covariance) looks right, but it's different from what shown in this PR.  Do you know what's going on?  Sorry for my inexperience on this.

pllim commented:
If the only difference is a `, dtype=int64`, maybe you can get away with using `+ELLIPSIS`

https://docs.python.org/3/library/doctest.html#doctest.ELLIPSIS

pllim commented:
Also the PR commit history here isn't clean anymore. Did you do this to rebase?

```
git fetch upstream main
git rebase upstream/main
```

I see a merge commit in your PR. That is not a good sign.

kbwestfall commented:
> Also the PR commit history here isn't clean anymore. Did you do this to rebase?
> 
> ```
> git fetch upstream main
> git rebase upstream/main
> ```
> 
> I see a merge commit in your PR. That is not a good sign.

A bit painful, but I think I successfully fixed my rebase issue.

pllim commented:
At least the commit history looks okay now but I didn't go through the diff again. Did you try ELIPSSIS for Windows?

kbwestfall commented:
> At least the commit history looks okay now but I didn't go through the diff again. Did you try ELIPSSIS for Windows?

Yes, thanks for the suggestion!  It worked.  I've also added some tests to increase coverage.  And I also added a change log file, but the "towncrier" CI test still fails.  Is there something else I should do to enable that test to pass?

pllim commented:
Yes, you need a What's New entry in `docs/whatsnew/7.0.rst`. :)

pllim commented:
Hmm. Maybe something got lost in the rebase. This is back.

`ModuleNotFoundError: Use of 'astropy.nddata.Covariance' requires 'scipy.sparse' module!`

kbwestfall commented:
> Yes, you need a What's New entry in `docs/whatsnew/7.0.rst`. :)

Thanks, and done!

On the test failure, this was because I added some more tests and forgot to decorate them with `pytest.mark.skipif`.  These are hopefully fixed now.

pllim commented on docs/whatsnew/7.0.rst:
```diff
@@ -123,6 +124,14 @@ value in ``Unit.to`` and have those arrays not be converted to Numpy arrays::
 
 Note that it is not yet possible to use ``Quantity`` with dask arrays directly.
 
+Constructing, accessing, and storing covariance matrices for N-dimensional data
```
```suggestion
.. _whatsnew_7_0_nddata_covariance:

Constructing, accessing, and storing covariance matrices for N-dimensional data
```

kbwestfall replied:
Thanks!

pllim commented:
I think the stakeholders here are:

* `nddata` maintainers
* spectroscopy group (do you have a lead over there?)

kbwestfall commented:
>     * spectroscopy group (do you have a lead over there?)

I expect @rosteen is the contact point.

pllim commented:
Okay, I think I tagged the people who should do a initial pass of final review. Please ping them (or request me to) if they do not respond. Also does not hurt to solicit feedback via astropy-dev mailing list or Astropy Slack if you have not already. Good luck!

rosteen commented:
I haven't had time to do a full in-depth review, but I didn't see anything concerning on a quick look through the code, and I definitely don't have any philosophical objections to adding this class.

mwcraig commented:
@kbwestfall -- apologies for the long delay in taking a look at this. 

I haven't spent much time thinking about covariance so I wanted to step back and understand what the underlying motivation/need is. There is a very high probability that I've made mistakes below...please let me know where I'm off.

1. I gather that the need for this arises from specutils, where there are correlated uncertainties and you need a way to express that.
2. Since `Spectrum1D` subclasses (via `NDCube`) from `NDData`, and the covariance expresses an aspect of the uncertainty, it makes sense to add it as a subclass of `NDUncertainty`.
3. It would be nice to have a way to serialize the covariance; your table methods do that, as do the FITS reader/writer. The ability to read/write covariance to FITS seems to be what prompted the original discussion.
4. The code itself implements what is in MaNGADAP but modified to fit within the NDData framework. 

Is that roughly the correct big picture?

kbwestfall commented:
Hi @mwcraig .  No worries, and thanks for having a look at the PR.  Yes, I agree that's a good summary of the picture.

mwcraig commented:
@kbwestfall ok, great -- I'm going to check out your branch and try it locally to get a feel for it. I think I have a couple questions but can work that out more easily if I give it a try 😀

mwcraig commented:
Thanks again for opening this -- I think it would add useful functionality to astropy that does not currently exist, and I think it is of broader interest than spectroscopy. I'm not 100% sure yet that nddata is the right place in astropy for it, but I'm about 90% there. 

I have a number of questions below; if you think it would be easier to have a conversation about them instead of going back and forth on GitHub, let me know -- I'm open later this week, and we can report back a summary to this thread.

Please keep in mind that my recent experience with correlation matrices is an hour or so of reading the AstroML book last night and about 30 minutes of playing around with something of the docstring example today 😬. I wouldn't start making any code changes until we've hashed out what the API looks like.

1. I found it a little confusing that whether `.var` existed depended on whether I'd called `variance()` or not. Could `.var` be made a property that is automatically calculated the first time the user accesses `.var`? In other words, it wouldn't be calculated at object creation but would automatically be calculated if/when the user acesses `.var`.
2. I also found it confusing that what I get from `Covariance.toarray()` depends on whether I've called `to_correlation` or `revert_correlation` or not. I also managed to get a `Covariance` object into a confused state by setting `.is_correlation` directly. What about having to properties, `.covariance_array` and `.correlation_array` that each produces the appropriate array? Then you can store things internally however you want and you need ot be explicit about which flavor you want.
3. On the I/O front, it seems like it would be easiest if the `Covariance` object converts cleanly to a `Table` and then `Table` I/O is used under the hood to write out/read in the files. In other words, if `Covariance.write` and `Covariance.read` could be wrappers around Table read/write. Then you could write them to FITS or ECSV or ASDF or whatever. It looks like the only change you would need to make for that to work is to add the variance to the metadata in the table produced by `to_tables` instead of return a variable vector and a table. 
4. Could this conceivably be integrated into `NDArithmeticMixin` down the road (not in this PR)? We allow for a correlation matrix, and having a `Covariance` object in nddata would make it easier to integrate that.

There were a couple other minor issues (like should `toarray` be `to_array` to match the other `to_...` things) but hashing through the stuff above seems like the higher priority.

rosteen commented:
@kbwestfall It looks like the cleanup bot will close this as stale soon. I don't remember when you said you'd have time to address Matt's comments, but maybe it would be good to do what the bot said and rebase just to make sure it doesn't get closed in the meantime.

mwcraig commented:
I can add the right label to keep it open also, though it needs a rebase anyway....

dhomeier commented:
Conflict basically needs the `whatsnew` to be moved to 7.1.rst, but I think that is difficult to effect in the web editor.

kbwestfall commented:
@rosteen @mwcraig @dhomeier Thanks for the heads up about this!  I had started working on some revisions based on @mwcraig's comments, but had to stop to deal with other things.  I'll get back to this within the next two weeks.

mwcraig commented:
Thanks for the update @kbwestfall -- I'm going to mark this `keep-open` for now just to make sure the bot doesn't close it.

kbwestfall commented:
Hi @mwcraig Thanks for these first comments, and sorry for the slow response!

>  I wouldn't start making any code changes until we've hashed out what the API looks like.

Despite this recommendation, I went ahead and made a bunch of changes :)

A major change is that the class now only stores the variance vector and the upper triangle of the correlation matrix.  This helps deal with some of the insightful issues you brought up below.  

>     1. I found it a little confusing that whether `.var` existed depended on whether I'd called `variance()` or not. Could `.var` be made a property that is automatically calculated the first time the user accesses `.var`? In other words, it wouldn't be calculated at object creation but would automatically be calculated if/when the user acesses `.var`.

I agree this was confusing.  I've changed the code so that variance is always available, and there's now a property method used to return it.

>     2. I also found it confusing that what I get from `Covariance.toarray()` depends on whether I've called `to_correlation` or `revert_correlation` or not. I also managed to get a `Covariance` object into a confused state by setting `.is_correlation` directly. What about having to properties, `.covariance_array` and `.correlation_array` that each produces the appropriate array? Then you can store things internally however you want and you need ot be explicit about which flavor you want.

Also agreed.  I changed this so this so that the `full` and `toarray` methods require you to select the type of matrix you want, instead of having to track which form the object is in (and potentially leading to the confused state you got it into).

>     3. On the I/O front, it seems like it would be easiest if the `Covariance` object converts cleanly to a `Table` and then `Table` I/O is used under the hood to write out/read in the files. In other words, if `Covariance.write` and `Covariance.read` could be wrappers around Table read/write. Then you could write them to FITS or ECSV or ASDF or whatever. It looks like the only change you would need to make for that to work is to add the variance to the metadata in the table produced by `to_tables` instead of return a variable vector and a table.

It would be useful to talk through this.  The issue with adding the variance vector to the table is that the number of variance entries and correlation entries are different.  I don't know how best to handle this in a single table.

>     4. Could this conceivably be integrated into `NDArithmeticMixin` down the road (not in this PR)? We allow for a correlation matrix, and having a `Covariance` object in nddata would make it easier to integrate that.

Another good thing to talk through.  Potentially, yes, but I would need to better understand the existing code.  Along the same lines, there are a few issues that I'm unsure about with having `Covariance` subclass from `NDUncertainty` (e.g., I foresee the `array` attribute of `NDUncertainty` being an issue for `Covariance`).  I've mostly done this as a convenience for interfacing with `specutils`, but there may be a better approach. 

> There were a couple other minor issues (like should `toarray` be `to_array` to match the other `to_...` things) but hashing through the stuff above seems like the higher priority.

On `toarray` vs. `to_array`, I agree, but I kept it as `toarray` because that's the same as used by `scipy.sparse` to convert from a sparse to a dense array, fwiw.

kbwestfall commented:
Hi all.  I have a few statements in the `docs/nddata/covariance.rst` file like this:
```
>>> np.allclose(covar.toarray(), _covar.toarray())
True
```

A few of the tests failed because of the difference between `np.True_` and `True`.  I just tried changing the above to:
```
>>> np.allclose(covar.toarray(), _covar.toarray())  # doctest: +ELLIPSIS
...True...
```
and that broke more things than it fixed.  I can keep trying different permutations, but, before I do that, does anyone know the right incantation?

neutrinoceros commented:
@kyleaoman whoops, looks like the one test where my previous suggestion doesn't work requires scipy and I didn't have it installed when I checked my solution earlier.

I see two possible solutions for it
```patch
diff --git a/docs/nddata/covariance.rst b/docs/nddata/covariance.rst
index b4efbba146..c9e695b3de 100644
--- a/docs/nddata/covariance.rst
+++ b/docs/nddata/covariance.rst
@@ -94,7 +94,7 @@ You can construct a covariance matrix based on samples from a distribution::
     >>> s = np.random.multivariate_normal(m, c, size=100000)
     >>> covar = Covariance.from_samples(s.T, cov_tol=0.1)
     >>> np.all(np.absolute(c - covar.toarray()) < 0.02)
-    True
+    np.True_
 
 Here, we have drawn samples from a known multivariate normal distribution with a
 given covariance between its 10 axes and checked the reconstruction of the
```

```patch
diff --git a/docs/nddata/covariance.rst b/docs/nddata/covariance.rst
index b4efbba146..2ca565ded3 100644
--- a/docs/nddata/covariance.rst
+++ b/docs/nddata/covariance.rst
@@ -93,7 +93,7 @@ You can construct a covariance matrix based on samples from a distribution::
     ...         + np.diag(np.full(10-2, 0.2, dtype=float), k=2))
     >>> s = np.random.multivariate_normal(m, c, size=100000)
     >>> covar = Covariance.from_samples(s.T, cov_tol=0.1)
-    >>> np.all(np.absolute(c - covar.toarray()) < 0.02)
+    >>> bool(np.all(np.absolute(c - covar.toarray()) < 0.02))
     True
 
 Here, we have drawn samples from a known multivariate normal distribution with a
```

(I have a slight preference for the latter)

kbwestfall commented:
> > (I have a slight preference for the latter)
> 
> Stylistically I'd prefer that, too, but from the point of being aware what `np.all` is returning here, the first option with `np.True_` might be better.

I agree with this, but my plan (assuming the new test passes) is to add some text to make sure the reader knows what `np.all` returns for numpy < 2 and numpy > 2.  Does that make sense?

neutrinoceros commented:
> what np.all returns for numpy < 2 and numpy > 2.

The return type didn't change with numpy 2.0, only the string representation (`repr`) of numpy booleans did.


```python
>>> import numpy as np
>>> np.__version__
'1.26.4'
>>> np.False_
False
>>> type(np.False_)
<class 'numpy.bool_'>
```
```python
>>> import numpy as np
>>> np.__version__
'2.2.1'
>>> np.False_
np.False_
>>> type(np.False_)
<class 'numpy.bool'>
```

(I note the name of the type also changed, though it's not affecting us in any way that I know of)

kbwestfall commented:
Good point, @neutrinoceros .  Thanks for the clarification!

mwcraig commented:
Thanks for the updates @kbwestfall -- I'll take a look at the revised PR tomorrow. @neutrinoceros thanks for your troubleshooting here!

kbwestfall commented:
Hi all.  I keep getting a pre-commit failure, but it passes for me locally.  Any idea what I could be doing wrong?

pllim commented:
pre-commit.ci autofix

neutrinoceros commented:
Just in case anyone is still confused by this: running `pre-commit` locally uses the versions of each hook as pinned on your branch, while `pre-commit.ci` first merges the target branch (`main` here) so they might go out of sync if the branching was done too long ago. In such a case, rebasing is your best option to get back to a healthy state (or merging `main` into your branch, though this has its own caveats).

mwcraig commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1269 @@
+        return n * 2 - self._rho.shape[0]
+
+    @property
+    def stored_nnz(self):
```
Also, why are they different values? At least I think they are different values....

mwcraig reviewed (changes requested):
Thanks for opening this @kbwestfall and apologies for the long delay in the review. THis PR is in really good shape, and the test coverage is very good. 

A few things in addition to the inline comments:
+ I'm hoping my relative unfamiliarity with using covariance made me a useful reviewer with no built in assumptions rather than a reviewer who raised a number of obvious points.
+ In the tests it would be great to have a single line or two of comment explaining what the test is checking for. 
+ My comments are almost entirely minor documentation, docstring, or naming suggestions. I did not, for better or for worse, attempt to verify that all of the index magic you do is correct. 
+ It would be nice to get closer to 100% test coverage on this as long as covering the remaining lines, which are mostly errors raised if the input is inappropriate, isn't too painful to do.

kbwestfall commented on astropy/nddata/covariance.py:
```diff
@@ -0,0 +1,1269 @@
+
+        # Convert the covariance matrix to a correlation matrix for storage
+        self._var, self._rho = Covariance.to_correlation(array)
+        # The correlation matrix is symmetric by definition (or it should be!),
```
I also put in a keyword argument that allows the user to skip the check.

kbwestfall reviewed (commented):
Thanks for the detailed review, @mwcraig!  I think I have addressed all the comments.  Some of the more major changes:

- I overhauled the main doc.  I moved most/all the examples there, and I tried to provide more complete coverage of the class capabilities.
- I changed the class so that it only holds the covariance data.  Breaking this into the variance vector and correlation matrix was a holdover from how we handled things in MaNGA, and I realized it was making the object overly complicated.
- I removed the explicit file I/O functions.  Instead, I only provide the to/from table methods, and then leave it to the user to use the table I/O methods to write the data to a file.  But I give an example of how to do this in the main doc.
- I improved the test coverage.

mwcraig commented on docs/nddata/covariance.rst:
```diff
@@ -0,0 +1,604 @@
+       [ 0. ,  0. ,  0. ,  0. ,  0. ,  0. ,  1.6,  6. , 16. , 10. ],
+       [ 0. ,  0. ,  0. ,  0. ,  0. ,  0. ,  0. ,  3. , 10. , 25. ]])
+
+Reordering and Subsets
```
I'm not quite sure where the right place in this section is to mention it but I think it is worth mentioning that the `Covariance` class has methods for doing these operations so that one does not have to do these (impressive) numpy gymnastics.

kbwestfall replied:
I added short note sections for each of these three sections pointing to the "utility" function descriptions at the end of the doc.

mwcraig commented on docs/nddata/covariance.rst:
```diff
@@ -0,0 +1,604 @@
+       [0.5, 1. , 0.5],
+       [0.2, 0.5, 1. ]])
+
+In N-dimensions
```
Same as the previous section, find a place to mention that `Covariance` provides methods for these.

kbwestfall replied:
Note added.

mwcraig reviewed (changes requested):
@kbwestfall -- I found this round of the documentation much clearer! There are a couple very minor suggestions and one request: Can you change the title "Uncertainties" on [this line](https://github.com/astropy/astropy/pull/16690/files#diff-891b27e6b46792d5376556163d6b03d21dc64040fd8c18c80abe63aad1f35bbbR214) to "Uncertainties and Covariance"?

That will make sure that it is more visible on the main `nddata` landing page under the "Using nddata" section of the documentation.

kbwestfall reviewed (commented):
Thanks for the valuable suggestions, @mwcraig!  I think I've addressed everything.

I also modified what was the `sub_matrix` after realizing that it was more general.  It's now called `match_to_data_slice`.

And I also realized, after looking at the Numpy documentation recently, that I could change the `bool()` calls in some of the examples to `print()`.

mwcraig reviewed (approved):
Thanks for doing this pull request @kbwestfall and for being patient as the reviews have dragged on!
